parts/linux/cloud-init/artifacts/cse_helpers.sh

```diff
@@ -683,4 +714,42 @@ removeKubeletFlag() {
     fi
 }
 
+oras_login_with_identity() {
+    local acr_url=$1
+    local client_id=$2
+    local tenant_id=$3
+
+    raw_access_token=$(curl "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fmanagement.azure.com%2F&client_id=$client_id" -H Metadata:true -s)
+    if [ -z "$raw_access_token" ]; then
+        echo "failed to retrieve kubelet token"
+        return $ERR_ORAS_PULL_UNAUTHORIZED
+    fi
+    ACCESS_TOKEN=$(echo "$raw_access_token" | jq -e .access_token -r 2>/dev/null)
+    if [ -z "$ACCESS_TOKEN" ] || [ "$ACCESS_TOKEN" = "null" ]; then
+        echo "failed to parse access token"
+        return $ERR_ORAS_PULL_UNAUTHORIZED
+    fi
+
+    raw_acr_token=$(curl -X POST "https://$acr_url/oauth2/exchange" \
+        -H "Content-Type: application/x-www-form-urlencoded" \
+        -d "grant_type=access_token&service=$acr_url&tenant=$tenant_id&access_token=$ACCESS_TOKEN" -s)
+    if [ -z "$raw_acr_token" ]; then
+        echo "failed to retrieve access token to acr '$acr_url', please check the kubelet identity access to acr"
+        return $ERR_ORAS_PULL_UNAUTHORIZED
+    fi
+    ACR_TOKEN=$(echo "$raw_acr_token" | jq -e .refresh_token -r 2>/dev/null)
+    if [ -z "$ACR_TOKEN" ] || [ "$ACR_TOKEN" = "null" ]; then
+        echo "failed to parse acr token"
+        return $ERR_ORAS_PULL_UNAUTHORIZED
+    fi
+
+    oras login $acr_url --identity-token $ACR_TOKEN --registry-config ${ORAS_REGISTRY_CONFIG_FILE}
+    ret=$?
```

Comment:
nit: don't usually need to capture exit codes like this when the command is running within the same shell as the script itself:

```suggestion
    if ! oras login $acr_url --identity-token $ACR_TOKEN --registry-config ${ORAS_REGISTRY_CONFIG_FILE}; then
        echo "failed to login to acr '$acr_url' with identity token"
        return $ERR_ORAS_PULL_UNAUTHORIZED
fi
```